### PR TITLE
refactor: Use events to test network logic

### LINKED ISSUE #2687: Unsubscribing from db events causes race condition
Found in #2686

### LINKED ISSUE #2699: Use events system to test networking logic
Refactor the events system so that it can be used to test networking events. After these changes are done we should be able to mock the networking logic in tests.

### PR DESCRIPTION
**Pull request by nasdf**
## Relevant issue(s)

Resolves #2699
Resolves #2687

## Description

This PR refactors the events package and updates the networking tests to use events.

I think there is a bit more clean up to be done, but the majority should be ready for review.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

`make test`

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: tests/integration/events/utils.go
Comment (fredcarle) [inline tests/integration/events/utils.go:93]: typo: invalid :)

File: events/bus.go
Comment (fredcarle) [reply events/bus.go:51]: suggestion: It seems like `any` could be changed to `struct{}`.

File: events/bus.go
Comment (fredcarle) [inline events/bus.go:68]: suggestion: Here too it seems like `any` could be changed to `struct{}`

**Review by fredcarle — COMMENTED** [review]
I need to head out but first pass looks really good. I love that you removed 1000 lines of code 🤘. I'll finish a more in depth review on Monday.

File: tests/integration/p2p.go
Comment (AndrewSisley) [inline tests/integration/p2p.go:416]: todo: Please document why you are doing this, it is unclear reading the code (same goes for the other similar block).  

todo: Please also document what `sourceSub` and `targetSub` do in the `setupReplicatorWaitSync` definition.

File: tests/integration/utils2.go
Comment (AndrewSisley) [inline tests/integration/utils2.go:661]: todo: Please document why `100`

File: tests/integration/utils2.go
Comment (AndrewSisley) [inline tests/integration/utils2.go:832]: todo: Please document why 100, and why you are subscribing

File: tests/integration/p2p.go
Comment (AndrewSisley) [inline tests/integration/p2p.go:418]: suggestion: I think `require` would be more appropriate here (and the other similar blocks in this file) - a failure here is a bug in the test, not a bug in the production code?  It would also probably be really confusing if the code continued past this point no?

todo: Please document why this assert/require is done

File: events/bus.go
Comment (AndrewSisley) [inline events/bus.go:52]: discussion: The scope of this mutex is quite large, and I don't think I agree with this change.  The previous solution was non-blocking and it didn't discard messages (the `default` block in `Bus`.`Publish` here).

I think it is very important that publishing events does not block 'core' database operations, particularly in user-space.  The rapid discarding of events if the buffer fills also far from ideal, and I worry that it will lead to some painful to figure out bug reports, particularly in the P2P space.

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
I think we will need to chat about this one - it might be worth separating the changes made outside of the `events` package from the rest - there are some very good changes in here that should not be delayed/confused with the events package implementation change.

Regarding the events package, it is now simpler, but it is also blocking (in user-space), and less reliable (it discards messages quite rapidly if the subscription buffer fills - the original buffer size was 100 + a 60 second timeout, it is now 10 items and no timeout) and as such I think I will require a bit of convincing in order to see this as a good change.

I do also disagree with the removal of the `Channel` interface, it was used to allow us to swap out implementations without having to change the calling code.  Whilst I think the original `simple` channel is better than the proposal for our use-cases, it still needs to be improved upon long-term, and different use-cases within the defra code base would likely benefit from different `Channel` implementations.

File: events/bus.go
Comment (fredcarle) [reply events/bus.go:52]: I don't think you can have non-blocking and also not discarding messages at the same time. You can delay blocking by having a certain size of channel but when it's full we can either block or discard. Unless we have a store to keep track of unsent messages but that is a whole other subject which probably shouldn't apply here. 

The previous solution was blocking as far as I can tell.

File: events/events.go
Comment (fredcarle) [inline events/events.go:38]: suggestion: Since we are in the `events` package, the `Event` suffix isn't necessary. We could change `events.ConnectEvent`, `events.PubSubEvent`, `events.UpdateEvent`, `events.MergeEvent` to `events.Connect`, `events.PubSub`, `events.Update`, `events.Merge`

File: events/bus.go
Comment (fredcarle) [reply events/bus.go:52]: We have 2 event bus in this PR. We have the internal event bus and the HTTP subscription event bus. The internal event bus will only ever block with the mutex on startup when it adds the internally defined and static subscribers. For the rest of its life, the mutex will never be blocking. Right now the `Publish` method is non-blocking on the channel and that is something we can debate. For internal message exchanges, blocking might actually be desirable.

The HTTP subscription bus, however, has a dynamic list of subscribers. This one will block more often to update the subscriber list. We do not want to block on the channel here that is certain. Since HTTP subscriptions touch user space, using timeouts in the `Publish` method would be appropriate.
```go
for id := range subscribers {
	go func(id int) {
		select {
		case b.subs[id].value <- msg:
			// published event
		case <-time.After(time.Second * 10):
			// channel full, timeout
		}
	}(id)
}
```

Comment (AndrewSisley) [reply events/bus.go:52]: > The previous solution was blocking as far as I can tell.

It was much less blocking - it was lock free, and the (buffered) `write` channel was consumed by a dedicated Go routine that fed inputs into the (buffered) subscribing channel. And for the write channel to block a subscriber channel would need to be full as well as the write channel.  Atm the write channel is hardcoded to size 0, and the subscribers 100.  It won't throw away data until a subscriber times out after 60 seconds.  We planned on making all of these settings configurable but never got round to it

The proposed solution blocks on every concurrent call regardless of configuration thanks to the quite broad mutex, and there is no way to configure around that (unlike a buffered write channel).  It will discard data as soon as the (size 10) subscriber buffer is filled - this is really not very tolerant at all.

The current solution is far from perfect, but I do think it is better than the proposal.

Comment (AndrewSisley) [reply events/bus.go:52]: > the mutex will never be blocking

The mutex will block the moment concurrent writers try to write to the bus.

File: events/bus.go
Comment (islamaliev) [inline events/bus.go:132]: todo: please either preallocate required space:
```go
subs := make([]*Subscription, len(b.subs))

b.mutex.RLock()
for i := range b.subs {
    subs[i] = b.subs[i]
}
```
or, even better, use `copy`:
```go
var subs []*Subscription

b.mutex.RLock()
copy(subs, b.subs)
```

File: events/bus_test.go
Comment (islamaliev) [inline events/bus_test.go:19]: suggestion: would be great to have more tests. Current one covers only the happy path and nothing else.
cases like these: subscribe twice to the same event, test unsubscribe, test if not subscribed-to event is not received, test close and others.

I'm sure it's mostly covered with integration tests, but would nice to have explicit unit tests for the event bus

File: events/events.go
Comment (islamaliev) [reply events/events.go:38]: that would be neat. But for this, following best go practices, would be better to make package name singular `event`.

File: tests/integration/events/utils.go
Comment (islamaliev) [inline tests/integration/events/utils.go:83]: question: shouldn't this be unsubscribed-from at some point?

File: events/events.go
Comment (fredcarle) [reply events/events.go]: @nasdf you changed to `event` but didn't do my suggestion. Did you just forget or did you decide you didn't like it? :)

Comment (nasdf) [reply events/events.go]: @fredcarle I'm considering moving the event definitions to their respective packages.

File: tests/integration/events/utils.go
Comment (nasdf) [reply tests/integration/events/utils.go]: node `Close` method takes care of unsubscribing.

File: tests/integration/state.go
Comment (nasdf) [reply tests/integration/state.go]: confirmed they are closed on node `Close`

File: events/bus.go
Comment (nasdf) [reply events/bus.go]: Implementation is now channel based like we discussed

File: tests/integration/utils2.go
Comment (nasdf) [reply tests/integration/utils2.go]: this is no longer relevant to the implementation

File: tests/integration/p2p.go
Comment (nasdf) [reply tests/integration/p2p.go]: moved to a helper func and added docs

File: events/events.go
Comment (fredcarle) [reply events/events.go]: Also a good option 👍

Comment (fredcarle) [reply events/events.go]: The downside of defining the events in their respective packages is that if might result in unwanted dependencies or even circular dependencies depending in which package it is defined. `MergeEvent` being defined in `db` or `net` for example.

File: net/process.go
Comment (fredcarle) [inline net/process.go:67]: praise: Make much more sense to have this here.

File: event/bus.go
Comment (AndrewSisley) [inline event/bus.go:56]: todo: I thought we agreed to keep this behind an interface?  Happy to resume that conversation if need be :)

File: http/client.go
Comment (AndrewSisley) [inline http/client.go:455]: thought: This can/should probably be implemented for http/cli at some point, half tempted to ask you to link this panic to a github issue.

File: internal/db/db.go
Comment (AndrewSisley) [inline internal/db/db.go:48]: todo: I think the default should be a much larger value, maybe 100,000 for now. 100,000 would still only be 1.6MB I think, and stuff will start blocking if the capacity is exceeded.

File: event/bus.go
Comment (AndrewSisley) [inline event/bus.go:56]: todo: Doesn't have to be in this PR, but the `OnFooAsync` funcs added in https://github.com/sourcenetwork/defradb/pull/2708 can now be removed IMO - they are dead code (unless we want to keep them to allow users to call them).

File: tests/integration/p2p.go
Comment (AndrewSisley) [inline tests/integration/p2p.go:518]: todo: Please document why these requires exist

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
Looks really good Keenan, I really like these changes.  Just got a couple of small todos/suggestions for you before merge :)

File: event/bus.go
Comment (AndrewSisley) [reply event/bus.go:56]: It already exists in the `events` package, this PR deleted it.  I think it would just need a rename, removal of the generic, and adding the `...string` param to `Subscribe`, otherwise it looks identical to the `*Bus` signature.

Comment (fredcarle) [reply event/bus.go:56]: It doesn't make sense to define the interface in the `event` package has it doesn't need/use it. We could define it in  `client` though for the purpose of the `client.DB` interface method `Events() Events`.

Comment (AndrewSisley) [reply event/bus.go:56]: If we are going to move it, I'd also suggest moving `Subscription` there too, and then the `event` package can become internal.

Otherwise, I do slightly disagree with you.  It is quite common for packages to define their public interfaces within themselves.  Especially given `event` is not really Defra specific, and could quite happily be moved outside of this repo for general use.

Comment (fredcarle) [reply event/bus.go:56]: Packages defining public interfaces that are not used internally is bad practice though. We should try to avoid it. The goal of an interface is to make something more flexible. Like if we want `db` to be able to use different event bus implementation. I'd argue that we are building the `Bus` specifically for our needs so I don't think it flexibility makes sense here. I much prefer the concrete type as it is now.

File: events/events.go
Comment (nasdf) [reply events/events.go]: Good point. I've kept them in the package but renamed as you suggested.

File: events/events.go
Comment (nasdf) [reply events/events.go]: I'd like to keep the event names a human readable string for debugging purposes.

Having the events in one package does have the issue you mentioned, but it also makes it much easier to know what events can be subscribed to. There's also the issue of importing an entire package just for an event type.

File: http/client.go
Comment (nasdf) [reply http/client.go:455]: I'm planning on adding this functionality in a follow up PR. One of my design goals was to allow the CLI to subscribe to and print the events as they happen.

File: event/bus.go
Comment (nasdf) [reply event/bus.go]: I'll try to remember to bring this up in the stand up

File: http/client.go
Comment (AndrewSisley) [reply http/client.go:455]: Sweet, sounds good :)

File: event/buffered_bus.go
Comment (AndrewSisley) [inline event/buffered_bus.go:26]: todo: Please add a compile time check to ensure (and document) that this type implements the `Bus` interface

File: event/bus.go
Comment (nasdf) [reply event/bus.go]: I added the interface because I will be using it to create a JavaScript event bus for the WASM build.

File: event/buffered_bus.go
Comment (nasdf) [reply event/buffered_bus.go:26]: The constructor method does this already. I usually choose one or the other not both.

Comment (AndrewSisley) [reply event/buffered_bus.go:26]: Fair enough :)  Fred might ask you to change the constructor to return the concrete though :grin:

File: events/events.go
Comment (islamaliev) [reply events/events.go]: but the package that wants to listed to a specific event usually depends on the package that fires that event as it needs to receive the package-event-specific object.

> I'd like to keep the event names a human readable string for debugging purposes

You can declare string enum:
```go
type EventName string
```
and in `db` package:
```go
const (
    // MergeCompleteEventName is the name of the database merge complete event.
    MergeCompleteEventName = events.EventName("db:merge-complete")
    // UpdateEventName is the name of the database update event.
    UpdateEventName = events.EventName("db:update")
    // ResultsEventName is the name of the database results event.
    ResultsEventName = events.EventName("db:results")
)
```

File: events/events.go
Comment (islamaliev) [reply events/events.go]: this is being discussed in another thread https://github.com/sourcenetwork/defradb/pull/2700#discussion_r1633762423 :)

File: event/buffered_bus.go
Comment (islamaliev) [inline event/buffered_bus.go:33]: question: what are you referring to by "simpleChannel"?

File: event/buffered_bus.go
Comment (islamaliev) [inline event/buffered_bus.go:38]: nitpick: sounds more like a NOTE, not a WARNING

File: event/buffered_bus.go
Comment (islamaliev) [inline event/buffered_bus.go:93]: todo: these 2 ops are not dome atomically and might close twice. Please use compare-and-swap:
```go
if !b.isClosed.CompareAndSwap(false, true) {
    return
}
...
```

Check please if you have similar occurrences.

**Review by islamaliev — CHANGES_REQUESTED** [review]
Great job, but there is at least one critical issue related to atomics that needs to be addressed before merge.

File: events/bus.go
Comment (islamaliev) [reply events/bus.go]: you marked the comment as "resolved" but I don't see how it was addressed. The same goes for https://github.com/sourcenetwork/defradb/pull/2700#discussion_r1633748528

Would be nice if for every comment you left a very short comment mentioning how you addressed it. That save reviewers' time. That's what we all tend to.

File: events/events.go
Comment (nasdf) [reply events/events.go]: > but the package that wants to listed to a specific event usually depends on the package that fires that event as it needs to receive the package-event-specific object.

This won't always be the case, and will cause problems since the `db` package is internal.

> You can declare string enum:

I've added an `event.Name` type.

File: event/buffered_bus.go
Comment (nasdf) [reply event/buffered_bus.go]: good catch!

File: event/buffered_bus.go
Comment (nasdf) [reply event/buffered_bus.go]: Fixed. It was a copy paste error.

File: events/bus.go
Comment (nasdf) [reply events/bus.go]: My bad. I thought I left a message. This should no longer be an issue with the implementation.

Comment (islamaliev) [reply events/bus.go]: > This should no longer be an issue with the implementation.

I don't quite understand what you mean. And I see no changes in the code.

Comment (nasdf) [reply events/bus.go]: The code referenced in the above comment has been replaced.

File: event/buffered_bus.go
Comment (islamaliev) [inline event/buffered_bus.go:98]: question: this is thread-safe only under the assumption that all the calls to `bufferedBus` are done by the same thread. I'm not sure this is the case. Is it?

Otherwise the following scenario is possible:
1. execution enters `Subscribe` method, `b.isClosed == false` so it proceeds but pauses just before `b.commandChannel <- subscribeCommand(sub)`
2. concurrently `Close` is being called, sets `b.isClosed` to `true`, pushes `b.commandChannel <- closeCommand{}`
3. `handleChannel` receives `closeCommand` and executes `close(b.commandChannel)`
4. now the thread from step 1 resumes, executes `b.commandChannel <- subscribeCommand(sub)` and causes a panic.

File: events/bus.go
Comment (islamaliev) [reply events/bus.go]: sorry, my bad. Didn't look at the last changes.

File: event/buffered_bus.go
Comment (shahzadlone) [inline event/buffered_bus.go:123]: question: Just pointing incase easy to test, that this path is not tested.

File: event/buffered_bus.go
Comment (shahzadlone) [inline event/buffered_bus.go:133]: question: Just pointing incase easy to test, that this path is not tested.

File: event/buffered_bus.go
Comment (shahzadlone) [inline event/buffered_bus.go:137]: question: Just pointing incase easy to test, that this path is not tested.

File: event/bus.go
Comment (shahzadlone) [inline event/bus.go:61]: nitpick: This path is not tested.

File: event/buffered_bus.go
Comment (AndrewSisley) [reply event/buffered_bus.go:113]: That's quite a broad mutex, please don't merge yet

Comment (nasdf) [reply event/buffered_bus.go:113]: It should only block when calling close.

Comment (AndrewSisley) [reply event/buffered_bus.go:113]: Okay, I think I understand, and there is no nicer way of handling this that I can spot off the top of my head.

todo: Can you please rename `mutex` to `closeLock` or similar, we really really don't want this to be used for anything else.

todo: Can please also document the lock on its declaration in `bufferedBus`, explaining that it is only locked whilst closing.

Comment (fredcarle) [reply event/buffered_bus.go:113]: @AndrewSisley can you explain what you mean by broad? We use the same technique in the memory store and I think it's really fine for this as well.

Comment (AndrewSisley) [reply event/buffered_bus.go:113]: > @AndrewSisley can you explain what you mean by broad? We use the same technique in the memory store and I think it's really fine for this as well.

The lock spans every public call to this struct, it can't really get any broader :)  If it only locks whilst closing that's fine, but it would be very damaging if that were to change (hence the documentation and name change requested).

File: event/buffered_bus_test.go
Comment (islamaliev) [inline event/buffered_bus_test.go:162]: praise: tests overall look good. Thanks for taking your time.

File: event/buffered_bus_test.go
Comment (nasdf) [reply event/buffered_bus_test.go]: I've decreased the delay. It's required because the publish is asynchronous.

File: event/buffered_bus_test.go
Comment (nasdf) [reply event/buffered_bus_test.go]: This call itself won't panic because it is simply adding a message to the command channel which is processed asynchronously.

File: event/buffered_bus_test.go
Comment (nasdf) [reply event/buffered_bus_test.go]: I've updated all the test names

File: event/buffered_bus_test.go
Comment (nasdf) [reply event/buffered_bus_test.go]: It was copy pasted, but I've fixed them all.

File: event/buffered_bus_test.go
Comment (nasdf) [reply event/buffered_bus_test.go]: Updated

**Review by islamaliev — APPROVED** [review]
LGTM! Thanks Keenan for the adjustments.

**Review by shahzadlone — APPROVED** [review]
Thanks for testing the requested paths.

LGTM

**Pull request merged**
